src/coreclr/tools/Common/Internal/Runtime/EETypeBuilderHelpers.cs

```diff
@@ -136,6 +141,23 @@ private static bool HasCriticalFinalizer(TypeDesc type)
             return false;
         }
 
+        private static bool IsTrackedReferenceWithFinalizer(TypeDesc type)
+        {
+            do
+            {
+                if (!type.HasFinalizer)
+                    return false;
+
+                if (((MetadataType)type).HasCustomAttribute("System.Runtime.InteropServices.ObjectiveC", "ObjectiveCTrackedTypeAttribute"))
```

Comment:
Look up of custom attributes is typically very expensive. I don't know if this was already mentioned, but if possible it would be better to set a bit on type load that can be checked. @MichalStrehovsky has this already been considered?

CoreCLR:
https://github.com/dotnet/runtime/blob/3855a99c9d88ed9dee53cd11c7533b53394d28d3/src/coreclr/vm/methodtablebuilder.cpp#L1956-L1981

Reply:
At runtime, a flag is used to lookup this state, so it should be fast:

https://github.com/dotnet/runtime/blob/b1eb66d5508e912d725e71d4fa2b52515d7ce8f3/src/coreclr/nativeaot/Common/src/Internal/Runtime/MethodTable.cs#L794-L798

At compile time, the presence of a finalizer is checked first, so this reduces the number of attribute checks. I added some instrumentation and compiled some unit tests to see how many classes have their attributes checked:

* src/tests/Interop/ObjectiveC/ObjectiveCMarshalAPI/ - **147**
* src/libraries/System.Runtime.InteropServices/tests/System.Runtime.InteropServices.UnitTests - **698**

Do you think that's a low enough number of checks to defer optimizing this to a follow up PR?

Reply:
> At runtime, a flag is used to lookup this state, so it should be fast:

Oh, I thought this was the regularly check at run-time. If this is just the initial check to set the flag, this is fine. Sorry.